### DISCUSSION TITLE
Laravel Collection keyBy Bug

### DISCUSSION THREAD

**Discussion by glupeksha** [laravel/framework#35279] [Q&A]
Hi, I was testing the below code in laravel tinker.
```
$c=collect([(object)['id'=>0,'name'=>'asd'],(object)['id'=>1,'name'=>'qwe']]);
$c->keyBy("id")
```
results for the above code is 
```
Illuminate\Support\Collection {#4317
     all: [
       {#4340
         +"id": 0,
         +"name": "asd",
       },
       {#4338
         +"id": 1,
         +"name": "qwe",
       },
     ],
   }
```
When I changed the id of the first element to a non zero value like the below code

```
$c=collect([(object)['id'=>3,'name'=>'asd'],(object)['id'=>1,'name'=>'qwe']]);
$c->keyBy("id")
```

It gives the intended results as below
```
Illuminate\Support\Collection {#4340
     all: [
       3 => {#4341
         +"id": 3,
         +"name": "asd",
       },
       1 => {#4339
         +"id": 1,
         +"name": "qwe",
       },
     ],
   }
```

Is there is a way to get a result like below when there is an element with value zero in laravel collection?
```
Illuminate\Support\Collection {#4340
     all: [
       0 => {#4341
         +"id": 0,
         +"name": "asd",
       },
       1 => {#4339
         +"id": 1,
         +"name": "qwe",
       },
     ],
   }
```

**Comment by kohlerdominik** [comment]
I did not test if it works, but you could try it with a closure like:
```php
$collection->keyBy(fn($item) => $item->id);
```

edit: if you're not using php 7.4 yet, you have to use the more expressive syntax for the closure:
```php
$collection->keyBy(function($item) {
    return $item->id;
});
```

**Comment by sachinkumar121** [comment]
In the case of an array key, Strings containing valid integers will be cast to the integer type. E.g. the key "8" will actually be stored under 8. On the other hand, "08" will not be cast, as it isn't a valid decimal integer.

Read this [reference link](https://stackoverflow.com/a/35522045/7182381) and another [link](https://stackoverflow.com/a/37167700/7182381)
